DEP: Design proposal for OnDelete Strategy (#879)
gardener/etcd-druid: docs/proposals/06-sts-ondelete-strategy.md

---
title: StatefulSet OnDelete Update Strategy
dep-number: 06
creation-date: 2024-09-20
status: implementable
authors:
- "@anveshreddy18"
reviewers:
- "etcd-druid-maintainers"
---

# DEP-06: Druid Controlled Pod Updates via StatefulSet OnDelete Strategy

## Summary

> This proposal recommends changing the StatefulSet update strategy used by etcd-druid from `RollingUpdate` to `OnDelete`. With `OnDelete`, the Kubernetes StatefulSet controller no longer automatically restarts pods when the pod template changes. Instead, a new dedicated controller in etcd-druid takes responsibility for deleting and recreating pods in a carefully chosen order that accounts for etcd member health and cluster role. The goal is to prevent unintended quorum loss during pod updates.

Reviewer: ```suggestion
This proposal recommends changing the StatefulSet update strategy used by etcd-druid from `RollingUpdate` to `OnDelete`. With `OnDelete`, the Kubernetes StatefulSet controller no longer automatically rollout the pods when the pod template changes. Instead, a new dedicated controller in etcd-druid takes responsibility for deleting and recreating pods in a carefully chosen order that accounts for etcd member health and cluster role. The goal is to prevent unintended quorum loss during pod updates.
```

> This proposal recommends changing the StatefulSet update strategy used by etcd-druid from `RollingUpdate` to `OnDelete`. With `OnDelete`, the Kubernetes StatefulSet controller no longer automatically restarts pods when the pod template changes. Instead, a new dedicated controller in etcd-druid takes responsibility for deleting and recreating pods in a carefully chosen order that accounts for etcd member health and cluster role. The goal is to prevent unintended quorum loss during pod updates.

Reviewer 2: ```suggestion
This proposal recommends supporting the `OnDelete` update strategy to be used by etcd-druid to update the StatefulSet pods backing the Etcd cluster members. With `OnDelete`, the Kubernetes StatefulSet controller no longer automatically restarts pods when the pod template changes. Instead, a new dedicated controller in etcd-druid takes responsibility for deleting and recreating pods in a carefully chosen order that accounts for etcd member health and cluster role. The goal is to prevent unintended quorum loss during spec updates to the Etcd cluster.
```

## Terminology

> - **OnDelete**: A StatefulSet update strategy where pods are only updated when they are explicitly deleted. The StatefulSet controller does not automatically roll pods on template changes.

Reviewer: ```suggestion
- **OnDelete**: A StatefulSet update strategy where pods are only updated when they are explicitly deleted. The StatefulSet controller does not automatically rollout the pods on template changes.
```

Consider a 3-member etcd cluster with pods `P-0`, `P-1`, and `P-2`. If `P-0` becomes unhealthy (due to network issues, node failure, or an internal error), the cluster still has quorum with `P-1` and `P-2`. Now, if a StatefulSet template update is triggered (for example, an image version bump), the StatefulSet controller starts rolling from `P-2`. It deletes `P-2` and waits for it to come back. During this window, only `P-1` is healthy and participating, which is below quorum (2 out of 3). The cluster experiences a quorum loss that could have been entirely avoided if the update had started with the already-unhealthy `P-0` instead.

The following diagram illustrates how the `RollingUpdate` strategy can lead to quorum loss in this scenario:

<div align="center">
<img src="assets/06-rolling-update-state-diagram.png" alt="RollingUpdate state diagram showing quorum loss when an unhealthy pod exists" width="500">
</div>

> The StatefulSet controller starts from Pod N (the highest ordinal), terminates it, and waits for the new pod to become ready. If the terminated pod is not the originally unhealthy one, quorum is lost with 2 members down.

Reviewer: Most likely it will be a transient quorum loss .. right as nth pod (highest ordinal) will eventually come-up  ?

Author: Yes, it'll be transient quorum loss. Will mention that clearly.

Reviewer: yeah, may be wording here can be improve, mention that "it can lead to transient quorum loss" or "temp. down time".

Author: yes I've updated it already, missed mentioning that in this conversation.

- Introduce a spec field on the Etcd custom resource that allows operators to choose between `RollingUpdate` and `OnDelete` strategies per cluster.
- Ensure that updates triggered by changes to the StatefulSet pod template (image versions, configuration, resource requests) are propagated to pods under druid's control.

### Non-Goals

- This proposal does not claim to prevent all forms of quorum loss. It reduces the likelihood of quorum loss caused by poorly-ordered voluntary pod updates.
- This proposal does not cover PVC resizing or volume replacement. The OnDelete strategy makes PVC resizing safer (see [Interaction with PVC Resizing](#interaction-with-pvc-resizing)), but the PVC resize flow itself is a separate feature.
- Optimizing the update process for clusters with more than 3 replicas (updating multiple pods concurrently while maintaining quorum) is left as future work.

## Proposal

### Update Strategy as an Etcd Spec Field

The update strategy will be exposed as a field on the Etcd custom resource:

```yaml
apiVersion: druid.gardener.cloud/v1alpha1
kind: Etcd
spec:
  updateStrategy: OnDelete  # or RollingUpdate
```

> - **Default value**: `OnDelete`

Reviewer 2: Shouldn't the default be `RollingUpdate`, to not break the current behaviour? Eventually it can be defaulted to `OnDelete`, maybe a few releases later, but not from the very beginning. Can you please mention this as a note maybe?

Author: If your concern is that `OnDelete` might need some testing in the landscapes before we go all-in, I agree with that - we can make the default `RollingUpdate` and move to `OnDelete` once we're confident about it.

But just curious, I didn't quite understand the below line
>to not break the current behaviour?

OnDelete vs RollingUpdate is an internal implementation change which doesn't affect the end state right? it only affects how it reaches that end state, from that perspective I don't think it makes much difference if we do it right away or after a few releases.

Pls correct me if you mean something else or I didn't understand it right. But I'll change that to `RollingUpdate` to allow selective testing of `OnDelete` on some seeds initially.

Reviewer 2: Since OnDelete is a new feature, we should ideally not make it the default behavior until it's well-tested in production. Hence, it should be an opt-in feature for now.

Author: Agreed! have made the changes my previous commit for this.

- **Valid values**: `RollingUpdate`, `OnDelete`

This is a per-cluster choice. Operators can set different strategies for different Etcd clusters. Changing the field on a live cluster is supported and triggers a seamless transition (see [Transitioning Between Strategies](#transitioning-between-strategies)).

> The decision to use a spec field instead of a feature gate is documented in [Decision Record 003](../decisions/003-ondelete-as-spec-field-not-feature-gate.md). The key reasons are: per-cluster control, no forced migration at maturity, and both strategies remaining available indefinitely.

Reviewer 2: I don't see this decision record file in the PR or in your fork. Please add it.

Author: oops, that's an internal backlog record that I misplaced here, do you think the decision record is useful in the context of this proposal too? I recorded it mainly for internal alignment. Pls lmk wdyt of this.

Reviewer 2: if it's internal, then it should not be mentioned in the DEP which is a public-facing doc.

Author: yeah but I meant whether making it public and referring in the DEP makes sense? But I think it's better to just remove this ref, will do that.

Reviewer 2: Yes, removing this ref is better.

### The OnDelete Controller

A new controller, separate from the existing Etcd reconciler, is responsible for managing pod updates when the OnDelete strategy is active. This controller watches StatefulSet resources and reconciles whenever the StatefulSet's `.status.updateRevision` changes, indicating that a pod template update needs to be propagated to pods.

**Why a separate controller instead of extending the StatefulSet component:**

> - The update process may span multiple reconciliation cycles (waiting for pods to come back, checking health, proceeding to the next pod). Embedding this in the StatefulSet component would make it a blocking operation for the entire Etcd reconciliation loop.

Reviewer 2: One could also argue that there are many long-running operations in the Etcd reconciliation loop today, but the pattern we've adopted is to actuate a change, requeue after some time, check whether that change has taken effect, [requeue again if not done yet], then proceed with the next step. Same logic would be applicable for the OnDelete logic too, so technically this could have been part of the statefulset component.

Does having two controllers lead to any possible race conditions? I just want to rule out such a possibility before we go ahead.

Reviewer 3: Indeed, what coordinates these two controllers when acting on the same StatefulSet concurrently? If the Etcd reconciler pushes a new STS pod template (new updateRevision) while the OnDelete controller is mid-rollout of the previous revision, what happens? I expect the OnDelete controller re-evaluates all pods against the latest revision, effectively restarting its procedure, but this should be stated explicitly, especially regarding potential update conflicts on the STS object itself.

Author: > Does having two controllers lead to any possible race conditions?

As STS controller allows only one strategy at the same time, it's purely an implementation specific on how we manage the state interally, a correct implementation of enabling and disabling ondelete controller only after STS updatestrategy is set to `OnDelete` and before STS updatestrategy is set to `RollingUpdate` respectively will be sufficient and won't lead to any race conditions.

> I expect the OnDelete controller re-evaluates all pods against the latest revision

Right, it evaluates basis on the latest updateRevision and now the previously updated pods will be again in-queue, so ondelete controller handles it well as it's stateless and always tries to be on the latest version + pick members in right order.

I'll address the conflict concern about two controllers acting simultaneously in the proposal.

- A separate controller can watch StatefulSet events independently and act asynchronously, which aligns with how the Kubernetes StatefulSet controller itself works.

**Controller predicate:** The controller only reconciles StatefulSets whose `spec.updateStrategy.type` is set to `OnDelete`. This means the controller is always registered in the controller manager but has zero overhead for clusters using `RollingUpdate`.

### Determining Whether a Pod Needs Updating

A pod is considered outdated if its `controller-revision-hash` label does not match the StatefulSet's `.status.updateRevision`. The `controller-revision-hash` is a Kubernetes-managed label on each pod that reflects the revision of the pod template it was created from.

> > Note: The `controller-revision-hash` is computed from the pod template spec only. Changes to `volumeClaimTemplates` do not affect this hash. See [Interaction with PVC Resizing](#interaction-with-pvc-resizing).

Reviewer 3: That's an internal implementation detail that may change. Unlikely, but possible, right? If that's indeed not a binding/public API contract, let's acknowledge that here with a sentence?

Author: > Unlikely, but possible, right?

Right, it can change. I'll mention that here.

**Step 2: Check for non-participating pods that need updating.** If any outdated pod is non-participating, select one for deletion using the following sub-priority:

| Priority | Container Status | Rationale |
|----------|-----------------|-----------|
| First | etcd process is dead (CrashLoopBackOff, Error, not running) | Already broken. Restarting it may fix the issue. |
| Second | etcd process is alive but readiness fails (quorum loss) | Process is healthy. Restarting it gets the new spec without making things worse. |

Delete the selected pod and requeue. Wait for it to come back before proceeding.

**Step 3: Ensure all updated pods are participating.** If any previously-updated pod is still not participating, requeue and wait. Do not update another participating pod until all already-updated pods have joined the quorum.

**Step 4: Update one participating pod.** At this point, all outdated pods are participating. Select one for deletion using this priority:

| Priority | Role | Rationale |
|----------|------|-----------|
| First | Follower | Deleting a follower does not trigger a leader election. |

> | Second | Leader | Deleting the leader triggers an election, so it is done last. |

Reviewer 2: I remember we had discussed long ago about manually moving the leader before updating the leader pod, so that we don't trigger a leader election at all, but gracefully move leadership and avoid even that intermittent leader election downtime. Can that be added here?

How do we move the leader? A very simple move-leader call to the leader etcd endpoint, possibly via an `EtcdOpsTask`. If you don't want scope creep, can you please add this as a future enhancement? Or if this is not much effort to implement, it could be included in this DEP and implementation or feature completeness.

Reviewer: I think we may not need it as I found on graceful deletion(shutdown) of a leader, etcd leader itself try to transfer the leadership before stoping the server.
https://github.com/etcd-io/etcd/blob/326d5a2e7765d1d918865d2c3897f0a27320db80/server/etcdserver/server.go#L1319-L1331

Author: Yes, the extra step of moving the leader is not needed.

Reviewer 2: Ok, if this is the case, then ignore my comment.

1. The operator sets `spec.updateStrategy: OnDelete` on the Etcd CR.
2. On the next Etcd reconciliation, the StatefulSet component updates the StatefulSet's `spec.updateStrategy` to `OnDelete`.
3. The OnDelete controller's predicate now matches this StatefulSet. If there are any outdated pods (from a previously in-progress RollingUpdate or from a new template change), the OnDelete controller picks up the work and starts updating pods in its health-aware order.
4. If no pods are outdated, the OnDelete controller simply watches for future StatefulSet template changes.

**Switching from OnDelete to RollingUpdate:**

1. The operator sets `spec.updateStrategy: RollingUpdate` on the Etcd CR.
2. On the next Etcd reconciliation, the StatefulSet component updates the StatefulSet's `spec.updateStrategy` to `RollingUpdate`.
3. The OnDelete controller's predicate no longer matches this StatefulSet. The Kubernetes StatefulSet controller resumes managing pod updates in its default ordinal order.
4. If there were outdated pods that the OnDelete controller had not yet processed, the StatefulSet controller picks them up and rolls them in the standard highest-to-lowest ordinal order.

> ### VPA and HVPA Interaction

Reviewer 2: Thanks for adding the section. There's no ned to mention HVPA, since it's no longer used.

Author: yea, will remove that. Thanks for flagging.

2. **VPA Admission Controller**: When the StatefulSet controller (or the OnDelete controller) recreates a pod, the admission webhook injects VPA-recommended resource values into the new pod's spec at creation time.

Since VPA does not modify the StatefulSet pod template, it does not trigger a new `updateRevision` and the OnDelete controller is not involved in VPA-driven resource updates. VPA continues to operate independently regardless of the update strategy.

### Interaction with PVC Resizing

The PVC resizing story ([etcd-druid#481](https://github.com/gardener/etcd-druid/issues/481)) benefits from `OnDelete` because it allows controlled per-pod volume replacement while maintaining quorum. However, the OnDelete controller and PVC resizing are independent features.

The OnDelete controller detects outdated pods by comparing `controller-revision-hash` labels. This hash is computed from the pod template spec only and does not include `volumeClaimTemplates`. This means that a change to `storageCapacity` or `storageClass` alone (without any pod template change) will not be detected by the OnDelete controller. The PVC resize flow must handle pod deletion independently in such cases.

> The details of the PVC resize flow (orphan-delete of the StatefulSet, per-pod PVC replacement, interaction with the OnDelete controller) will be covered in a separate proposal.

Reviewer: lets not mention the extra details which is not decided yet, may be just mention the issue no. https://github.com/gardener/etcd-druid/issues/481 to follow.

Author: > just mention the issue

Sure, will do that.

> not mention the extra details 

Which extra details do you mean here? These -> `(orphan-delete of the StatefulSet, per-pod PVC replacement, interaction with the OnDelete controller)` ? I can remove them from this proposal but added just to give some context.

Reviewer 2: I think mentioning about PVC resize just confuses readers since this DEP is mainly about OnDelete and has nothing to do with PVC resizing (although we had previously discussed both together, but then decided/realised that they are not dependent on each other and can be decoupled).

Author: I just wanted to address the PVC resize usecase also as it should not look like we've not thought about PVC resizing in the context of this DEP.

I removed the explicit details and mentioned the issue as Ishan suggested, I hope that's alright.

Reviewer: ok, thanks .. I will check

### Metrics

The OnDelete controller exposes the following metrics:

> - `etcd_druid_ondelete_update_duration_seconds`: Time from the first detection of an `updateRevision` change to the completion of all pod updates. Labeled by etcd cluster name.

Reviewer 2: ```suggestion
- `etcddruid_ondelete_update_duration_seconds`: Time from the first detection of an `updateRevision` change to the completion of all pod updates. Labeled by etcd cluster name.
```

This is the current convention, but if it seems wrong, we can discuss changing it in the current code too, during implementation.

> - `etcd_druid_ondelete_reconcile_cycles_total`: Number of reconciliation cycles required to complete a full pod update. Labeled by etcd cluster name.

Reviewer 2: Same as above

### Future Scope

- **Backup-restore container health in update ordering**: The current design does not consider the health of the backup-restore sidecar container when deciding which pod to update next. The rationale is that backup-restore health does not affect quorum, and prioritizing it could lead to unnecessary leader elections (for example, if a pod with an unhealthy backup-restore happens to be the leader). If future operational experience shows value in considering backup-restore health as a secondary sorting criterion, the priority order can be extended. The following state diagram illustrates what a backup-restore-aware ordering would look like:

  <div align="center">

>   <img src="assets/06-OnDelete-StateDiagram-With-Etcdbr-Health.png" alt="OnDelete state diagram with backup-restore health awareness (future scope)" width="700">

Reviewer 2: I feel we don't need this diagram, it would just confuse readers of this DEP, since you state we don't worry about etcdbr health in this DEP.

Author: Okay, I'll remove the embedding from the proposal but keep the diagram in assets for future use. I hope that's fine.

**Why a separate controller instead of extending the StatefulSet component:**

The StatefulSet component can potentially do the update via the existing Etcd reconciliation loop. We choose a separate controller because:

- **Pod-by-pod waits do not block the reconciler.** A full update can span many reconciliation cycles, since each pod must come back and rejoin the quorum before the next one is touched. A separate controller progresses asynchronously and does not tie up a reconciler thread per cluster for the duration of an update, which is unnecessary and could lead to scalability issues if many clusters are updating simultaneously.
- **Separation of concerns.** The Etcd reconciler ensures the StatefulSet spec matches the desired state among other things; the OnDelete controller propagates that spec to existing pods in a health-aware order. Keeping these in separate controllers makes each loop simpler to reason about.

**Coordination with the Etcd reconciler.**

The Etcd reconciler writes the StatefulSet spec; the OnDelete controller only deletes pods. They don't contend on the same field, so there is no update conflict on the StatefulSet object itself.

> The OnDelete controller is stateless: i.e every reconciliation re-reads the current `.status.updateRevision` and the current pod set. If the Etcd reconciler pushes a new pod template while OnDelete is mid-rollout of a prior revision, the next reconciliation observes the new `updateRevision`, treats every pod whose `controller-revision-hash` no longer matches as outdated (including ones already updated against the prior revision), and continues the procedure from there. No explicit handover between controllers is needed.

Reviewer: ```suggestion
The OnDelete controller is stateless: i.e every reconciliation re-reads the current StatefulSet's `.status.updateRevision` and the current pod set. If the Etcd reconciler pushes a new pod template while OnDelete is mid-rollout of a prior revision, the next reconciliation observes the new `updateRevision`, treats every pod whose `controller-revision-hash` no longer matches as outdated (including ones already updated against the prior revision), and continues the procedure from there. No explicit handover between controllers is needed.
```

The strategy switch itself is coordinated by sequencing: when `spec.updateStrategy` changes on the Etcd CR, the StatefulSet component updates the StatefulSet's `spec.updateStrategy` first, and the OnDelete controller's predicate (`spec.updateStrategy.type == OnDelete`) ensures it engages only when the StatefulSet is in the matching mode. The Kubernetes StatefulSet controller and the OnDelete controller therefore never act on pod updates concurrently.

**Controller predicate:** The controller only reconciles StatefulSets whose `spec.updateStrategy.type` is set to `OnDelete`. This means the controller is always registered in the controller manager but has zero overhead for clusters using `RollingUpdate`.

### Determining Whether a Pod Needs Updating

A pod is considered outdated if its `controller-revision-hash` label does not match the StatefulSet's `.status.updateRevision`. The `controller-revision-hash` is a Kubernetes-managed label on each pod that reflects the revision of the pod template it was created from.

> > **Note:** The `controller-revision-hash` is an internal contract of the Kubernetes StatefulSet controller rather than a documented public API. Changes to its semantics are unlikely but possible; if Kubernetes alters how this label is populated, the comparison performed here would need to be revisited.

Reviewer 4: I think we shouldn't mention the string `controller-revision-hash` in the DEP. `appsv1.StatefulSetRevisionLabel` is part of the public API. So, I think as long as we're relying on this, we're safe.

Author: Thanks for finding this, it'll definitely help to use this const over the string but (maybe by wording is wrong there) the main concern here is that the `controller-revision-hash` is populated in the Pod's `.metadata.labels.controller-revision-hash` and that's not guaranteed to be always populated this way, so I meant that could change and if that happens the comparison performed would need to be revisited.

Reviewer 4: I don't quite understand the concern here. If Kubernetes changes how the revision hash is calculated, our logic should remain the same, if I'm not wrong? Since we don't rely on the format of the hash or any internal detail and just do string comparisons.

Author: As discussed offline: The confusion stemmed from the understanding that what if kubernetes alters the hash computation mechanism, but what the proposal meant was more about what if the `controller-revision-hash` is not populated in the pod.metadata.labels or if there's a change in how the sts determines if pod's under it are upto date or not, as that contract is not public since the string is part of the label, so it can change in the future - if that happens we might want to change how the ondelete controller determines which pods are upto date and which are outdated. I'll enhance the doc a bit to clear the confusion.

Author: done, pls check :)

### Pod Update Procedure

The controller processes one pod per reconciliation cycle. After updating a pod, it requeues and waits for the pod to come back before proceeding to the next one. This ensures that the cluster is never in a state where more than one pod is being updated at the same time due to the controller's own actions.

The procedure in each reconciliation cycle:

**Step 1: Check if all pods are up to date.** Compare each pod's `controller-revision-hash` with the StatefulSet's `.status.updateRevision`. If all match, the update is complete.

**Step 2: Check for non-participating pods that need updating.** If any outdated pod is non-participating, select one for deletion using the following sub-priority:

| Priority | Container Status | Rationale |
|----------|-----------------|-----------|

> | First | etcd process is dead (CrashLoopBackOff, Error, not running) | Already broken. Restarting it may fix the issue. |

Reviewer 4: Could we be explicit about the statuses under which this priority is assigned? Does "not running" include cases like the container in "Waiting" for example? Also https://github.com/gardener/etcd-druid/pull/879#discussion_r3233063344 applies

Author: Addressed, PTAL :)

> | Second | etcd process is alive but readiness fails (quorum loss) | Process is healthy. Restarting it gets the new spec without making things worse. |

Reviewer 4: Nit:
```suggestion
| First | etcd process is dead (CrashLoopBackOff, Error, not running) | Already broken. Restarting the pod may fix the issue. |
| Second | etcd process is alive but readiness fails (quorum loss) | Process is healthy. Restarting the pod gets the new spec without making things worse. |
```
Since the restart is happening at the pod level even though the trigger is at the container level.

Author: done, pls check :)

**Step 3: Ensure all updated pods are participating.** If any previously-updated pod is still not participating, requeue and wait. Do not update another participating pod until all already-updated pods have joined the quorum.

**Step 4: Update one participating pod.** At this point, all outdated pods are participating. Select one for deletion using this priority:

| Priority | Role | Rationale |
|----------|------|-----------|
| First | Follower | Deleting a follower does not trigger a leader election. |
| Second | Leader | Deleting the leader may trigger a brief leader election, so it is done last. |

The controller determines member roles by reading the member lease's `holderIdentity` field, which contains the member ID and role.

> **Note:** When the leader pod is deleted with a normal grace period, the etcd server attempts a graceful [leadership transfer](https://github.com/etcd-io/etcd/blob/326d5a2e7765d1d918865d2c3897f0a27320db80/server/etcdserver/server.go#L1319-L1331) before shutting down, so in practice the cluster does not observe a full leader-election downtime. An explicit move-leader call from the controller before updating a leader is therefore not needed.

> Delete the selected pod and requeue. Wait for it to come back before proceeding to the next.

Reviewer 4: I think `Wait for it to come back before proceeding to the next.` could be a separate step at the start of the reconciliation procedure. Since the controller is stateless, I wonder what the approach would be to make sure we pick the pod we just deleted recently to wait for. Or is it a general `AllMembersReady` kind of wait?

Author: It's more like `AreAllUpdatedMembersReady` condition that gets checked, the wording is just to keep the story narrow, but if you think it's confusing, then I can improve on that.

Reviewer 4: I think we can be more explicit in the wording. My understanding was that it would wait for all members to be ready before the requeue, but if it was mentioned explicitly, it would be a lot clear.

Author: yeah, I'll improve the wording here a bit. Thanks.

Author: the step 3 in the update procedure covers this now, pls check.

2. **Eviction with `unhealthyPodEvictionPolicy: IfHealthyBudget`**: Allows eviction of unhealthy pods when enough healthy pods exist. Handles the common case of one unhealthy pod, but still deadlocks when two or more pods are unhealthy.

3. **Eviction with `unhealthyPodEvictionPolicy: AlwaysAllow`**: Allows eviction of all unhealthy pods regardless of budget. Solves the deadlock but weakens PDB protection against external actors (VPA, node drain, cluster autoscaler) that could concurrently evict unhealthy pods.

4. **Direct Delete**: Simple and does not interact with the PDB at all.

The recommended approach is direct `Delete` for the following reasons:

- The OnDelete controller is a conscious actor that makes health-aware decisions about which pod to delete and when. Unlike the StatefulSet controller (which blindly follows ordinal order), the OnDelete controller checks participation status, role, and liveness before every deletion. The risk of accidentally causing quorum loss is minimal because the controller explicitly verifies cluster state before acting.

> - The theoretical race condition (an external actor disrupts a pod between the controller's health check and its delete call) is possible but rare in practice, and it exists equally with the current `RollingUpdate` strategy, which also uses direct pod deletions internally.

Reviewer 4: I'm not sure I understand the benefits of `Delete` over the `Eviction` API w.r.t these 2 points. I think these 2 points are applicable regardless of the deletion approach though.

Author: While the first point is generic, the second one differs for `Delete` vs `Eviction` and that point is actually a limitation and not a benefit. Why it matters is that if we use `Eviction` calls, then our eviction call in this scenario would be rejected whereas if we use the `Delete` call, it goes through, so there's a difference and this point talks about that mainly.

If you'd like me to reorganize the points to clear any confusion, I can do that.

Reviewer 4: I think it would be better to reorganize. WDYT?

The first point could be removed and the second point could be modified to include the fact that "if we use Eviction calls, then our eviction call in this scenario would be rejected whereas if we use the Delete call" since I feel it makes it clear about how the `Delete` and `Eviction` approaches are different. Since the second point is actually a limitation compared to the `Eviction` approach, I'd consider moving it to a separate section as this section deals with why `Delete` is the recommended approach (I'm assuming we're just focusing on the pros here).

Author: Agreed, the first point is generic and second is actually a con, so we can reorganise this block of information for better readability and also expand on the second point as you mentioned.

Author: done, pls check :)

### Safeguarding Etcd Pods from Voluntary Disruptions

Two mechanisms protect etcd pods from being disrupted by external actors:

1. **PodDisruptionBudget (PDB)**: For a 3-member etcd cluster, `minAvailable` is set to 2, allowing at most 1 pod to be evicted at a time. For single-node clusters, `minAvailable` is 0 (PDB provides no protection). The PDB protects against evictions from VPA, node drain, and similar actors. The OnDelete controller bypasses the PDB via direct delete calls because it makes its own health-aware decisions.

2. **Cluster autoscaler `safe-to-evict` annotation**: The `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation prevents the cluster autoscaler from evicting etcd pods during node scale-down.

### Single-Node Clusters

For single-node etcd clusters, the OnDelete controller follows the same procedure but with a simplification: there is only one pod, so there is no ordering decision to make. The controller deletes the single outdated pod and waits for it to come back.

> Before deleting the single pod, the controller should trigger an on-demand snapshot (if backup is configured) to ensure that the latest data is persisted. This is because the single-node cluster has no redundancy: if the pod restart fails and a restoration is needed, the snapshot provides a recent recovery point.

Reviewer 4: In this case, I think we need another step in the beginning of the reconciliation loop to requeue if an on-demand snapshot is in progress. WDYT? Also I think we need to consider what happens if the on-demand snapshot fails - do we proceed with the update, are retries to be setup, etc.

Author: Actually now that I think more about it - it makes more sense to do it as part of Statefulset component `PreSync` independent of `OnDelete` or `RollingUpdate` flow as both do the same thing in a single-node case and both cases benefits from having a full snapshot taken before updates - but we can do that for image updates for now like how we did for etcd 3.5 upgrade, as doing for other pod template spec changes would require us to compute the hash diff, we can do that as well if the benefit to cost is justified. 

I think this decoupling will also allow the ondelete controller to not worry about snapshotting & stuff as well. Wdyt?

Reviewer 4: Yes, I'm also in favour of decoupling the ondelete controller from this. The snapshot on presync before update for non-HA could be taken care of as a separate issue. Also, given the member removal DEP (DEP-07), we might change how non-HA updates happen, and thus we might not need a presync snapshot.

Author: yeah, I'll raise an issue & PR for that separately to decouple them.

**Switching from OnDelete to RollingUpdate:**

1. The operator sets `spec.updateStrategy: RollingUpdate` on the Etcd CR.
2. On the next Etcd reconciliation, the StatefulSet component updates the StatefulSet's `spec.updateStrategy` to `RollingUpdate`.
3. The OnDelete controller's predicate no longer matches this StatefulSet. The Kubernetes StatefulSet controller resumes managing pod updates in its default ordinal order.
4. If there were outdated pods that the OnDelete controller had not yet processed, the StatefulSet controller picks them up and rolls them in the standard highest-to-lowest ordinal order.

Because the Kubernetes StatefulSet controller also consults `controller-revision-hash` and `.status.updateRevision` to decide which pods to roll, pods already updated by the OnDelete controller are not re-rolled - the StatefulSet controller only processes the remaining outdated ones.

### VPA Interaction

> The Vertical Pod Autoscaler (VPA) does not modify the StatefulSet spec directly. VPA operates through two independent mechanisms:

Reviewer 4: Nit:
```suggestion
The Vertical Pod Autoscaler (VPA) in `Recreate` or `InPlaceOrRecreate` mode does not modify the StatefulSet spec directly. VPA operates through two independent mechanisms:
```
Just to avoid generalization based on defaults specific to Gardener.

Author: done, pls check :)

1. **VPA Updater**: Evicts pods that don't match the recommended resource values. This eviction respects the PDB.
2. **VPA Admission Controller**: When the StatefulSet controller (or the OnDelete controller) recreates a pod, the admission webhook injects VPA-recommended resource values into the new pod's spec at creation time.

Since VPA does not modify the StatefulSet pod template, it does not trigger a new `updateRevision` and the OnDelete controller is not involved in VPA-driven resource updates. VPA continues to operate independently regardless of the update strategy.

### Interaction with PVC Resizing

The PVC resizing story ([etcd-druid#481](https://github.com/gardener/etcd-druid/issues/481)) benefits from `OnDelete` because it allows controlled per-pod volume replacement while maintaining quorum. However, the OnDelete controller and PVC resizing are independent features, and the PVC resize flow will be covered in a separate proposal.

> The OnDelete controller detects outdated pods by comparing `controller-revision-hash` labels. This hash is computed from the pod template spec only and does not include `volumeClaimTemplates`. This means that a change to `storageCapacity` or `storageClass` alone (without any pod template change) will not be detected by the OnDelete controller. The PVC resize flow must handle pod deletion independently in such cases.

Reviewer 4: If this is the case, if pod deletion has to be handled independent of the OnDelete logic, I think the point above this would not be valid. Since it is the responsibility of the PVC resizing flow to ensure quorum.

Author: From offline discussion with Rinish: Agreed, it's better to remove any dependency between OnDelete & PVC resizing for now and just keeping the reference to how these two relate together is enough for now. I'll be making the changes.

Author: done, pls check :)

## Alternatives

### RollingUpdate with `maxUnavailable`

Kubernetes supports a `maxUnavailable` field on the StatefulSet's `RollingUpdate` strategy ([StatefulSet documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#maximum-unavailable-pods)). Setting `maxUnavailable: 1` ensures that only one pod is unavailable during the update process.

This approach has a key limitation: the StatefulSet controller waits for any unavailable pod to recover before proceeding. If a pod is already unhealthy and the update itself might fix it (for example, a configuration change or image bump), the controller is stuck waiting. The `OnDelete` strategy does not have this limitation because it can proactively delete the unhealthy pod first, potentially resolving the issue.

Additionally, the `maxUnavailable` field does not allow control over the order of pod updates. The StatefulSet controller still follows the ordinal order (highest to lowest), which may not align with the cluster's health state.

> For these reasons, `OnDelete` with a health-aware controller provides a more robust solution for etcd clusters where quorum safety is critical.

Reviewer 4: I think both `OnDelete` and `maxUnavailable` are equivalent in terms of quorum safety. A fair point could be that with `OnDelete` we minimize the number of leadership transfers due to awareness about the member status.

Author: > I think both OnDelete and maxUnavailable are equivalent in terms of quorum safety.

Yeah, you're right.

> with OnDelete we minimize the number of leadership transfers due to awareness about the member status.

Not only that, the point above this also talks about update making progress vs getting stuck with ondelete vs maxUnavailable respectively.

---
title: StatefulSet OnDelete Strategy
dep-number: 06
creation-date: 18th Sep'2023
status: implementable
authors:
- "@anveshreddy18"
reviewers:
- "etcd-druid-maintainers"
---

# DEP-06: Druid Controlled Updates for Etcd Pods

## Summary

This Druid Enhancement Proposal (DEP) proposes to change the updateStrategy of the Etcd StatefulSet deployed by `etcd-druid` operator from `RollingUpdate` to `OnDelete`.

This is done to better orchestrate the rolling of the pods in a way to avoid any unintended quorum loss scenarios introduced because of the current rolling update strategy.

## Terminology

- **[OnDelete](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies)**: A statefulset update strategy that updates the pods only when they are deleted.

> - **[RollingUpdate](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#rolling-updates)**: A statefulset update strategy that updates the pods one by one in a rolling fashion.

Reviewer: ```suggestion
- **[RollingUpdate](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#rolling-updates)**: A statefulset update strategy that updates the pods from the largest ordinal to the smallest in a rolling fashion.
```

- **Quorum**: The minimum number of members required to form a consensus in a distributed system like etcd.

> - **Leader**: The member of the etcd cluster that is responsible for handling all the client requests and replicating the data to the followers.

Reviewer: Leader doesn't handle all the client request, technically it only handles all write req and read req can be served by any voting member.
```suggestion
- **Leader**: The member of the etcd cluster that is responsible for handling all the write requests and replicating the data to its followers.
```

- **Follower**: The member of the etcd cluster that replicates the data from the leader.

> - **Learner**: The member of the etcd cluster that does not participate in the quorum and is used for learning the data from the leader.

Reviewer: ```suggestion
- **Learner**: The non-voting member of the etcd cluster that does not participate in the leader-election and is used for learning the data from the leader.
```

For a single node etcd cluster, both the `RollingUpdate` and `OnDelete` strategies work the same way as there is only a single pod to roll in case of an update to the statefulset. However, for a multi-node etcd cluster, the `OnDelete` strategy can be used to avoid unintended quorum loss scenarios by controlling the order of pod updation.

## Goals

- Provide a way to avoid unintended quorum loss scenarios introduced because of the current rolling update strategy.

> - This design document proposes the introduction of a feature gate in the `etcd-druid` operator. The feature gate is intended to provide users with the ability to select the `OnDelete` update strategy for the StatefulSet deployed by the operator.

Reviewer: why a feature-gate ? why not configure it via Etcd api 🤔 by which operator can choose which update strategy they want to use, and later we can set default strategy to "OnDelete"

- This proposal does not cover the usecase of autonomous clusters where the etcd cluster is deployed as static pods instead of a StatefulSet. The proposal is limited to the StatefulSet deployment of the etcd cluster as it only changes the update strategy of the StatefulSet.

## Proposal

To avoid the beforementioned scenarios of unintended quorum loss, one can make use of the `OnDelete` updateStrategy provided by the statefulset. This essentially disables the automatic rollouts of pods upon statefulset pod template spec updates, and instead rolls the pods only when they are deleted.

This gives `etcd-druid` a way to control the order of pod updation by taking into account the health and role of the pods (follower or leader).

### Things to consider

#### Health Assessment Of Containers

The overall health of an etcd pod is determined by the status of its constituent containers. An etcd pod primarily consists of two containers:

> 1. **etcd container**: Runs the etcd process. The health of the etcd container is evaluated using a [readiness probe](https://kubernetes.io/docs/concepts/configuration/liveness-readiness-startup-probes/#readiness-probe). This probe checks if the etcd container is ready to handle traffic.

Reviewer: ```suggestion
1. **etcd container**: Runs the etcd process. The health of the etcd container is evaluated using a [readiness probe](https://kubernetes.io/docs/concepts/configuration/liveness-readiness-startup-probes/#readiness-probe). This probe checks if the etcd container is ready to serve incoming traffic.
```

> 2. **backup-restore container**: Operates as a sidecar to the etcd container. No readiness probe is defined for the backup-restore container. Its health is not a primary concern post-initialization since it does not impact the quorum.

Reviewer: ```suggestion
2. **backup-restore container**: Operates as a sidecar to the etcd container. No readiness probe is defined for the backup-restore container. Its health is not a primary concern post-initialization of etcd member since it does not impact the etcd cluster.
```

> In the context of the `OnDelete` update strategy, the health assessment is focused solely on the etcd container. This is because, post-initialization, the backup sidecar's health does not impact the quorum. There is a possibility of the sidecar container becoming unhealthy which may falsely indicate that the pod can be deleted to update it, but has the risk of losing quorum. `backup-restore` container's health is not a significant concern as the etcd container is the primary entity responsible for handling traffic and maintaining the quorum. The primary objective of the `OnDelete` strategy is to prevent quorum loss, not to ensure the `backup-restore` sidecar container's health.

Reviewer: ```suggestion
In the context of the `OnDelete` update strategy, the health assessment is focused solely on the etcd container. This is because, post-initialization, the backup sidecar's health does not impact the etcd's quorum. There is a possibility of the sidecar container becoming unhealthy which may falsely indicate that the pod can be deleted to update it, but has the risk of losing quorum. `backup-restore` container's health is not a significant concern as the etcd container is the primary entity responsible for handling traffic and maintaining the quorum. The primary objective of the `OnDelete` strategy is to prevent potential permanent or transient quorum loss, not to ensure the `backup-restore` sidecar container's health.
```

Understanding the health of individual containers within the pod is crucial for the update process to avoid quorum loss situations. The protection mechanisms for etcd pods vary depending on whether it is a single-node or multi-node cluster, which is essential for comprehending the update strategy.

#### Safeguarding Etcd Pods from Voluntary Disruptions

Two primary mechanisms are in place to protect etcd pods from being [evicted](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api) due to voluntary disruptions:

1. The [PodDisruptionBudget](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) (PDB) configured for the etcd pods.
2. The [safe-to-evict](https://kubernetes.io/docs/reference/labels-annotations-taints/#cluster-autoscaler-kubernetes-io-safe-to-evict) annotation applied by the Cluster Autoscaler (CA) to the etcd pods to prevent cluster scale-down.

> In a single-node etcd scenario, the PDB `.spec.minAvailable` is set to `0`, but the pod carries a CA annotation `cluster-autoscaler.kubernetes.io/safe-to-evict` set to `false`. This configuration prevents the pod from being evicted by the cluster autoscaler during voluntary disruptions. However, the pod remains susceptible to eviction by the Vertical Pod Autoscaler (VPA) or eviction API calls, thus allowing the `OnDelete` process to evict the pod if necessary.

Reviewer: ```suggestion
In a single-node etcd scenario, the PDB `.spec.minAvailable` is set to `0`, but the pod carries a CA annotation `cluster-autoscaler.kubernetes.io/safe-to-evict` set to `false`. This configuration prevents the pod from being evicted by the cluster autoscaler during voluntary disruptions. However, the pod remains susceptible to eviction by the Vertical Pod Autoscaler (VPA) as it uses eviction API call, thus allowing the `OnDelete` process to evict the pod if necessary.
```

> For a three-member etcd cluster, the PDB `minAvailable` is set to `2` to ensure the availability of at least two pods at any given time to maintain quorum. This configuration permits the `OnDelete` process to evict only a single pod at any time from a three-member etcd cluster.

Reviewer: > This configuration permits the `OnDelete` process to evict only a single pod at any time from a three-member etcd cluster

unlike PDB "OnDelete" doesn't uses eviction API ... then why you have mentioned this ....I didn't get this part.
